### PR TITLE
Migrate header search to a Stimulus controller (basic version)

### PR DESCRIPTION
**Pull request by lb-**
- Removes the jQuery slide animation so content will be instantly replaced (we do not appear to have the animation for content replacement in the search results within modals and other similar places so I assumed it was not something worth rebuilding)
- Removes the `autofocus` behaviour on the search fields as this is not helpful for screen readers / keyboard control
- Sets the `aria-busy` attribute on the results container while loading
- Includes support for `window.header` if provided
- Base implementation for #9950

---

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments? Chrome 109, Firefox 109, Safari 16.1 on macOS 13.0

### CODE REVIEW

File: wagtail/admin/templates/wagtailadmin/shared/header.html [hunk 1]
```diff
@@ -34,7 +34,15 @@ <h1 class="w-header__title" id="header-title">
...
             {% if search_url %}
-                <form class="col search-form" action="{% url search_url %}{% if query_parameters %}?{{ query_parameters }}{% endif %}" method="get" novalidate role="search">
+                <form
+                    class="col search-form"
+                    action="{% url search_url %}{% if query_parameters %}?{{ query_parameters }}{% endif %}"
+                    method="get"
+                    novalidate
+                    role="search"
+                    data-controller="w-search"
+                    data-action="{% verbatim %}change->w-search#search input->w-search#search{% endverbatim %}"
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/header.html:44]: Note that the `verbatim` is required here due to the HTML auto-formatter treating `>` inside data attributes as the end tag of the HTML element. Loveth - we can use this approach to work around the problem for now.

We will likely need to raise a bug to track this.
cc @thibaudcolas  / @Lovelyfin00

**Comment by lb-** [comment]
Thinking that `w-inject` may be better for this controller - as it's not just for search but rather injection of html from a response.

File: wagtail/admin/templates/wagtailadmin/shared/header.html (hunk 1 above)
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/header.html:44]: This has been resolved now. I'll rebase this and remove the verbatim usage.

File: client/src/controllers/SearchController.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  HTMLFormElement | HTMLInputElement
+> {
+  static targets = ['input'];
+
+  static values = {
+    loading: { default: false, type: Boolean },
+    pageParam: { default: 'p', type: String },
+    queryParam: { default: 'q', type: String },
+    results: { default: '#results', type: String },
+    url: { default: '', type: String },
```
Comment (lb-) [inline client/src/controllers/SearchController.ts:25]: `src` may be a better value name.

File: client/src/controllers/SearchController.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+ */
+export class SearchController extends Controller<
+  HTMLFormElement | HTMLInputElement
+> {
+  static targets = ['input'];
+
+  static values = {
+    loading: { default: false, type: Boolean },
+    pageParam: { default: 'p', type: String },
+    queryParam: { default: 'q', type: String },
```
Comment (lb-) [inline client/src/controllers/SearchController.ts:23]: A simpler approach might be to pull this from the input's name.

File: client/src/controllers/SearchController.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+ * input can be the controlled element or the containing form.
+ */
+export class SearchController extends Controller<
+  HTMLFormElement | HTMLInputElement
+> {
+  static targets = ['input'];
+
+  static values = {
+    loading: { default: false, type: Boolean },
+    pageParam: { default: 'p', type: String },
```
Comment (lb-) [inline client/src/controllers/SearchController.ts:22]: It might be better to provide the Params to clear in the event Params.

File: client/src/controllers/SearchController.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    const iconContainer = (
+      this.hasInputTarget ? this.inputTarget : this.element
+    ).parentElement;
+
+    this.icon = iconContainer?.querySelector('use') || null;
+    this.iconName = this.icon?.getAttribute('href') || null;
+    this.loadingValue = false;
+
+    // set up debounced search action method
+    this.search = debounce(this.runSearch.bind(this), this.waitValue);
```
Comment (lb-) [inline client/src/controllers/SearchController.ts:91]:
```suggestion
    this.update = debounce(this.search.bind(this), this.waitValue);
```
Update may be a better method name

File: client/src/controllers/SearchController.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    }
+  }
+
+  /**
+   * Perform a search based on a single input query, and only if that query's value
+   * differs from the current matching URL param. Once complete, update the URL param.
+   * Additionally, clear the 'page' param in the URL if present.
+   */
+  runSearch() {
+    const pageParam = this.pageParamValue;
```
Comment (lb-) [inline client/src/controllers/SearchController.ts:163]: Instead of being tied to the input, it probably makes more sense for this to serialise the form, which should contain a `p` reset, merge this with the url values (so we don't overwrite existing ones) and then apply the request based on the form. 

This means p may get requested also. Might be ok.

**Comment by lb-** [comment]
@thibaudcolas  I have cleaned this up a bunch more - would love to get your initial pass on feedback for this when you get the chance.

File: client/src/controllers/SwapController.ts [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    const iconContainer = (
+      this.hasInputTarget ? this.inputTarget : this.element
+    ).parentElement;
+
+    this.iconElement = iconContainer?.querySelector('use') || null;
+    this.iconValue = this.iconElement?.getAttribute('href') || '';
+    this.loadingValue = false;
+
+    // set up debounced update method
+    this.update = debounce(this.swapLocation.bind(this), this.waitValue);
```
Comment (lb-) [inline client/src/controllers/SwapController.ts:88]: This should be simpler. Just wrap the existing method without the rename. 

```suggestion
    this.swapLocation = debounce(this.swapLocation.bind(this), this.waitValue);
```

Then .. use that in the html. 

```
data-action="change->w-swap#swapLocation"
```

**Comment by lb-** [comment]
@thibaudcolas  I have updated this implementation to set aria-busy on the target container while 'loading', I opted out of setting aria-live="polite" on the elements though as it appears we have inconsistent header implementations so we end up with confusing screen reader output.

Happy to fix this also but I think it may increase the scope of this PR, happy to raise a new issue once this is merged.

### Example (with aria-live)

**match results** - works great as that uses a `h2` on the results with role=alert.

Screenshot 2023-02-04 at 1 09 20 pm

It also reads 'busy' if the screen reader has 'time'.

**no results** - works great as that uses a `h2` on the results also

Screenshot 2023-02-04 at 1 16 47 pm

**clear search** - does not work well at all, it reads the entire content (no h2)

https://github.com/wagtail/wagtail/blob/main/wagtail/images/templates/wagtailimages/images/results.html#L5

Screenshot 2023-02-04 at 1 17 39 pm

**Comment by laymonage** [comment]
@lb- Could you rebase this please, as I think there might be some adjustments needed after #10487. I'd be keen to review this PR within the next few days. Thanks!

**Comment by lb-** [comment]
@laymonage should be good to review now, I've rebased, re-read the code and smoke tested again.

File: client/src/controllers/SwapController.ts [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+    // only do the query if it has changed for trimmed queries
+    // for example - " " === "" and "first word " ==== "first word"
+    if (currentQuery.trim() === newQuery.trim()) return;
+
+    // Update search query param ('q') to the new value or remove if empty
+    if (newQuery) {
+      searchParams.set(queryParam, newQuery);
+    } else {
+      searchParams.delete(queryParam);
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:215]: I might've missed it but it seems we don't have a test case that clears the input after it was previously entered

File: client/src/controllers/SwapController.ts [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    return targeElement as HTMLElement;
+  }
+
+  /**
+   * Toggle the visual spinner icon if available and ensure content about
+   * to be replaced is flagged as busy.
+   */
+  loadingValueChanged(isLoading: boolean) {
+    if (isLoading) {
+      this.targetElement?.setAttribute('aria-busy', 'true');
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:123]: I couldn't get this to be read by VoiceOver, but I might have used/configured it incorrectly

**Review by laymonage — COMMENTED** [review]
This is looking very good, thanks! I've tested this on Chrome 114, Firefox 114, and Safari 16.5.1 on macOS 13.4.1. All seems to work well.

Just some comments on the code, but I think functionality-wise it looks solid to me.

Looks like the old config with `window.headerSearch` still works. Are you planning to remove the usages in a separate PR? What about the release notes?

File: client/src/controllers/SwapController.ts [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+        this.dispatch('success', {
+          cancelable: false,
+          detail: { results },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: targetElement as HTMLInputElement,
+        });
+      })
+      .catch((error) => {
+        if (error.name === 'AbortError') return;
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:177]: Great find! Thank you

**Comment by lb-** [comment]
Thanks for the review, I'll get to the individual items soon. 

In regards to progress on the full header search migration, I have intentionally kept this scope small so that it could be reviewed faster. 

I have some additional work somewhere that worked through the next stages of the Stimulus adoption but I felt it was going to cause the PR to get too large and I wanted feedback on the approach taken.

For now, this PR alone should not need upgrade considerations, but a future PR might. I expect we would keep support for the global up until 6.0 though, even if upcoming PRs remove our own admin usage of it.

File: client/src/controllers/SwapController.ts [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    this.loadingValue = false;
+
+    // set up debounced update method
+    this.update = debounce(this.swapLocation.bind(this), this.waitValue);
+  }
+
+  getTarget(targetValue = this.targetValue) {
+    const targeElement = document.querySelector(targetValue);
+
+    const foundTarget = targeElement && targeElement instanceof HTMLElement;
```
Comment (lb-) [inline client/src/controllers/SwapController.ts:99]:
```suggestion
    const foundTarget = targetElement && targetElement instanceof HTMLElement;
```

File: client/src/controllers/SwapController.ts [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+          !hasValidUrlValue && 'Cannot find valid src URL value',
+          !foundTarget &&
+            `Cannot find valid target element at "${targetValue}"`,
+        ]
+          .filter(Boolean)
+          .join(', '),
+      );
+    }
+
+    return targeElement as HTMLElement;
```
Comment (lb-) [inline client/src/controllers/SwapController.ts:114]:
```suggestion
    return targetElement as HTMLElement;
```

File: client/src/controllers/SwapController.ts [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+        }
+        return response.text();
+      })
+      .then((results) => {
+        const targetElement = this.targetElement as HTMLElement;
+        targetElement.innerHTML = results;
+
+        this.dispatch('success', {
+          cancelable: false,
+          detail: { results },
```
Comment (lb-) [inline client/src/controllers/SwapController.ts:171]:
```suggestion
          detail: { requestUrl, results },
```

[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:171]: Great idea.

File: client/src/entrypoints/admin/wagtailadmin.js [hunk 14]
```diff
@@ -36,3 +36,11 @@ window.addEventListener('load', () => {
   initAnchoredPanels();
   initMinimap();
 });
+
+/**
+ * When page search results are successful, reinitialise widgets
+ * that could be inside the newly injected DOM.
+ */
+window.addEventListener('w-swap:success', () => {
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/entrypoints/admin/wagtailadmin.js:44]: Yes, once the tooltips are fully migrated (will need another PR on top of #10620 though), then this line here will no longer be needed.

File: client/src/controllers/SwapController.ts [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    const targeElement = document.querySelector(targetValue);
+
+    const foundTarget = targeElement && targeElement instanceof HTMLElement;
+    const hasValidUrlValue = !!this.srcValue;
+
+    if (!foundTarget || !hasValidUrlValue) {
+      throw new Error(
+        [
+          !hasValidUrlValue && 'Cannot find valid src URL value',
+          !foundTarget &&
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:106]: I will refactor this to be a bit less esoteric, I was probably trying to be a bit too fancy here and it's hard to follow the logic.

File: client/src/controllers/SwapController.ts [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  /**
+   * Perform a search based on a single input query, and only if that query's value
+   * differs from the current matching URL param. Once complete, update the URL param.
+   * Additionally, clear the 'page' param in the URL if present, can be overridden
+   * via action params if needed.
+   */
+  swapLocation({ params = {} }: { params?: { clear?: string } }) {
+    const clearParams = (params?.clear || 'p').split(' ');
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:199]: Done, good call.

File: client/src/controllers/SwapController.ts (hunk 8 above)
Comment (lb-) [reply client/src/controllers/SwapController.ts:214]: Good call, have added now.

**Comment by lb-** [comment]
@laymonage ok should be good for another review.

I can do another smoke test for the aria-busy, I do recall (been a while) I had to slow down the voice read rate but I did validate that the screen reader correctly read out that the content had been updated.

As for the legacy `window.headerSearch` support, I have added more support here with an `afterLoad` method, so that if any `window.headerSearch` is used AND its `termInput` finds an input without a controller attached to its form. Then it will set up the right data attributes automatically.

I still think it's not worth doing the upgrade notes as part of this PR though, I hope to get another PR done after this one which will remove ALL `window.headerSearch` inline script usage and replace with the data attribute approaches. However, the goal is to have this PR be able to go to release if that next PR is not yet done.

File: client/src/controllers/SwapController.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts]: This should be cleaned up now.

File: client/src/controllers/SwapController.ts (hunk 9 above)
Comment (lb-) [reply client/src/controllers/SwapController.ts:170]: I can do another smoke test for the aria-busy, I do recall (been a while) I had to slow down the voice read rate but I did validate that the screen reader correctly read out that the content had been updated.

Comment (lb-) [reply client/src/controllers/SwapController.ts:170]: Maybe also slow down your network when running locally, aria-busy will only trigger for <300ms when doing a local API request.

File: client/src/controllers/SwapController.ts [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+          cancelable: false,
+          detail: { error, requestUrl },
+          // Stimulus dispatch target element type issue https://github.com/hotwired/stimulus/issues/642
+          target: this.targetElement as HTMLInputElement,
+        });
+        // eslint-disable-next-line no-console
+        console.error(`Error fetching ${requestUrl}`, error);
+      })
+      .finally(() => {
+        this.loadingValue = false;
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:237]: I think we should ensure that the promise is for the latest request, as otherwise this would set the `loadingValue` to `false` when there is a request that is aborted after a new one has started and is in progress.

I was able to reproduce this issue by throttling my network to be really slow via devtools (e.g. 1 kBps, or maybe 5kBps is enough): there is an ongoing request, but because a previous request is aborted, it sets the `loadingValue` to `false`.

The following seems to fix this:

```suggestion
        if (signal === this.abortController?.signal) {
          this.loadingValue = false;
        }
```

File: client/src/controllers/SwapController.ts [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+const getGlobalHeaderSearchOptions = (): {
+  targetOutput?: string;
+  termInput?: string;
+  url?: string;
+} => window.headerSearch || {};
+
+/**
+ * Allow for an element to trigger an async query that will
+ * patch the results into a results DOM container. The query
+ * input can be the controlled element or the containing form.
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:25]: I think we should mention its ability to update the location as well

File: client/src/controllers/SwapController.ts [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+    }
+  }
+
+  /**
+   * Abort any existing requests & set up new abort controller, then fetch and patch
+   * in the HTML results, dispatching an event and handling any clean up.
+   * Cancel any in progress results request using the AbortController so that
+   * a faster response does not replace an in flight request.
+   */
+  async request(param: string | CustomEvent<{ url: string }>) {
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:184]: This seems like a useful method that can be used as the Stimulus action when you just need to patch the DOM without updating the location. Perhaps rename this to `patch()` (or `swap()`, but that might be confusing), as this method does more than just firing a request?

**Review by laymonage — COMMENTED** [review]
A few more things that I should've noticed earlier (sorry!), and some comments on the naming before we get this in (as it'll be harder to change later). Once those are resolved, I think this should be ready to be merged 😄

File: client/src/controllers/SwapController.ts [hunk 20]
```diff
@@ -0,0 +1,288 @@
...
+
+  connect() {
+    // set up values that may be set with the global header approach
+    const formContainer = this.hasInputTarget
+      ? this.inputTarget.form
+      : this.element;
+    const { targetOutput, url: src } = getGlobalHeaderSearchOptions();
+    this.srcValue =
+      src || this.srcValue || formContainer?.getAttribute('action') || '';
+    this.targetValue = targetOutput || this.targetValue;
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:119]: Yep. Agreed, I am very much hoping I'll be able to pull this headerSearch fallback out, I might just change the behaviour now to put this all in the afterLoad code. That way there is no reading of the global in our controller.

File: client/src/controllers/SwapController.ts [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+    this.iconValue = this.iconElement?.getAttribute('href') || '';
+
+    // set up initial loading state (if set originally in the HTML)
+    this.loadingValue = false;
+
+    // set up debounced update method
+    this.updateLocation = debounce(
+      this.swapLocation.bind(this),
+      this.waitValue,
+    );
```
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:138]: We actually need both, in the next PR there is a use-case for a non-debounced version. 

Additionally you may not want to debounce for a one off (e.g. on html load) call. 

I could call these `updateLocationLazy` / `updateLocation`. 

Alternatively I have a POC code that pulls out the debounce to the data-action. 

You would do something like `data-action="input->w-swap#update.debounce`


Option 3 is we refactor and use a Param. 
 `data-action="input->w-swap#update.debounce" data-w-swap-delay-param="50"`.

I did do option 3 originally but ran into some Typescript headaches but I can give it another go.

[the remark this answers is not in the document]
Comment (laymonage) [reply client/src/controllers/SwapController.ts:138]: `updateLocationLazy` / `updateLocation` sounds good. Other options can also be e.g. "delayed" instead of "lazy", or make it a prefix e.g. `updateLocation` / `delayedUpdateLocation`

> Option 3 is we refactor and use a Param.
data-action="input->w-swap#update.debounce" data-w-swap-delay-param="50".
> I did do option 3 originally but ran into some Typescript headaches but I can give it another go.

Don't we already have and use the `waitValue` to configure the debounce? What is the difference?

[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts:138]: we need a way to override it on a per call basis, I will try something out.

File: client/src/controllers/SwapController.ts (hunk 9 above)
Comment (lb-) [reply client/src/controllers/SwapController.ts:170]: Nice digging, thanks. Keen to maybe re-look at this in the future though.

**Comment by lb-** [comment]
@laymonage  ready for another round, thank you for taking the time to review this thoroughly. I feel like this is a really fine tuned approach now, let me know if anything stands out.

File: client/src/controllers/SwapController.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (lb-) [reply client/src/controllers/SwapController.ts]: Just an update on this comment. I've now changed the PR to be fully reliant on this afterLoad code. This means no html changes in this PR now, it's all done in one place. 

I've removed the depreciation warning on the function, as now we must have this call made until a new PR comes in to adopt the data attributes approach. 

I've started this future work but I'll avoid new PRs until we are ready and confident with this approach, and we've merged this PR and the next search PR.

File: client/src/controllers/SwapController.ts [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+  targetOutput?: string;
+  termInput?: string;
+  url?: string;
+} => window.headerSearch || {};
+
+/**
+ * Allow for an element to trigger an async query that will
+ * patch the results into a results DOM container. The query
+ * input can be the controlled element or the containing form.
+ * It supports tha ability to update the URL with the query
```
Comment (laymonage) [inline client/src/controllers/SwapController.ts:26]: 😄 

```suggestion
 * It supports the ability to update the URL with the query
```

**Review by laymonage — APPROVED** [review]
Looking very solid here, thanks for continuing to work on this! So happy to see all the controller code covered by tests. One small typo but I think that can be fixed as you merge this. (I can also merge if you want.)

Thanks again!

File: client/src/controllers/SwapController.ts (hunk 22 above)
Comment (lb-) [reply client/src/controllers/SwapController.ts:26]: Thanks. Oh man, I think I need a spell checker on commit changes or something.

**Comment by lb-** [comment]
Thank you @laymonage  - I know it's a big job to review some of these PRs.

I am excited to see us start to have a baseline of shared, reusable JS behaviours built with a similar approach... and unit tested :D
